fix: reject modifications that terminate flow scopes required by activation

tsedekey opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->
This PR prevents invalid process instance modifications where a sub-process element is terminated while simultaneously activating a child element within it. This previously led to inconsistent process states where the activated element had no valid flow scope.

The solution follows [Option B](https://github.com/camunda/camunda/issues/37079#issuecomment-3254205191): such conflicting modification requests are now explicitly rejected with a clear error message. This ensures the engine remains in a consistent and predictable state.

A failing test was added to demonstrate the issue, and now verifies that the invalid scenario is correctly handled.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #37079

tsedekey committed:
test: reproduce bug in sub-process modification

Adds a unit test in ModifyProcessInstanceTest to simulate a process
instance modification
where a sub-process is terminated and a child element inside it is
activated.

tsedekey committed:
fix: activation without flow scope is rejected

test: remove placeholder test

refactor: address PR comments

test: update rejection type and reason

refactor: improve efficiency of validation logic

refactor: suggestions from code review

berkaycanbc reviewed (changes requested):
Thanks @tsedekey for working on this! 🚀 

⌛This one is open for a long time. Please keep an eye on PRs that are open for a long time because they can become outdated easily or miss to be in the next release. Please ask for another review if the original reviewer is not available.

⭐ Test cases are quite nicely covering different cases and the implementation looks good semantically.

❌ There is one case that is not working with the current implementation and needs to be handled; nested flow scopes. The test case below reproduces the original error `"!INVALID_STATE (Expected flow scope instance with key '<id>' to be present in state but not found.)"`:

```java
@Test
  public void shouldRejectActivationWhenFlowScopeIsTerminatedNested() {
    // given
    ENGINE
        .deployment()
        .withXmlResource(
            Bpmn.createExecutableProcess(PROCESS_ID)
                .startEvent()
                .subProcess(
                    "subprocess",
                    sp ->
                        sp.embeddedSubProcess()
                            .startEvent()
                            .serviceTask("A", a -> a.zeebeJobType("A"))
                            .subProcess(
                                "subprocess-nested",
                                spn ->
                                    spn.embeddedSubProcess()
                                        .startEvent()
                                        .serviceTask("B", a -> a.zeebeJobType("B"))
                                        .endEvent("sub-end-nested"))
                            .endEvent("sub-end"))
                .endEvent()
                .done())
        .deploy();

    final var processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID).create();
    final var subProcessElementInstance =
        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
            .withProcessInstanceKey(processInstanceKey)
            .withElementType(BpmnElementType.SUB_PROCESS)
            .withElementId("subprocess")
            .getFirst();

    // when
    final var rejection =
        ENGINE
            .processInstance()
            .withInstanceKey(processInstanceKey)
            .modification()
            .activateElement("sub-end-nested")
            .terminateElement(subProcessElementInstance.getKey())
            .expectRejection()
            .modify();

    // then
    assertThat(rejection)
        .describedAs("Expect that activation is rejected when flow scope is terminated")
        .hasRejectionType(RejectionType.INVALID_ARGUMENT)
        .hasRejectionReason(
            """
              Expected to modify instance of process '%s' but it contains one or more activate instructions \
              for elements whose required flow scope instance is also being terminated: element 'sub-end-nested' requires flow scope instance '%s' \
              which is being terminated. Please provide a valid ancestor scope key for the activation or avoid terminating the required flow scope."""
                .formatted(PROCESS_ID, subProcessElementInstance.getKey()));
  }
```

The implementation needs to be adjusted to handle nested flow scopes.

berkaycanbc commented:
@tsedekey any updates on this?

tsedekey commented:
> @tsedekey any updates on this?

@berkaycanbc first of all, thank you for your patience with this PR, I’ve taken your point about not leaving PRs open for too long, and I’ll make a conscious effort to avoid that in future.

I’ve now addressed all your comments, and I really appreciate your suggestion to add the missing test — that was a crucial case, and I’ve updated the code accordingly to ensure that it passes as well.

Thanks again for your thoughtful review. Looking forward to your re-review! 🙏

tsedekey commented:
Hi @berkaycanbc, thanks again for taking the time to sync with me the other day to walk through the issue and suggest improvements to the validation logic. Your approach to handling activations of elements whose flow scopes are being terminated — especially in the deeply nested cases — really helped clarify things for me. I now have a much better understanding of how flow scopes work and why your solution is more efficient.

I’ve implemented your suggested validation method, and I’ve also added a unit test that specifically covers the scenario where a three level nested element is activated while its parent flow scope (several levels up) is being terminated. The test correctly rejects the modification as expected. 🚀 🙌

berkaycanbc commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java:
```diff
@@ -674,6 +687,149 @@ private boolean isFlowScopeOfElement(
+    return Either.left(new Rejection(RejectionType.INVALID_ARGUMENT, reason));
+  }
+
+  private Either<Rejection, ?> validateFlowScopeNotTerminatedForActivation(
```
🔧 This method is now unused, should be removed.

berkaycanbc commented on zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationModifyProcessor.java:
```diff
@@ -674,6 +687,149 @@ private boolean isFlowScopeOfElement(
+    for (final var terminatedInstanceKey : terminatedInstanceKeys) {
+      final var terminatedInstance = elementInstanceState.getInstance(terminatedInstanceKey);
+      if (terminatedInstance == null) {
+        continue;
```
```suggestion
        // this should be already validated in validateElementInstanceExists
        continue;
```

berkaycanbc reviewed (approved):
Thanks @tsedekey for accepting the challenge and working on this bug! 🚀 

👍 I've suggested some last minor changes we need to tackle before merging this PR but they do not require another review.

✅ Pre-approved.

tsedekey committed:
test: update related test

tsedekey commented:
/backport
